### PR TITLE
Guardrails against bad config

### PR DESCRIPTION
**Pull request by j00bar**
Ensure that all defined backends:

- have valid routes
- have origins inside the cluster
- don't override reserved Turnpike prefixes and stay within trusted URL space
- require auth outside of specific URL prefixes

And write tests to prove it.

### CODE REVIEW

File: nginx/build_config.py [hunk 1]
```diff
@@ -13,7 +13,38 @@
...
 
 # Do not include leading or trailing slashes in these routes
-FORBIDDEN_ROUTES = ["", "saml", "auth", "_nginx"]
+PROTECTED_ROUTES = ["saml", "auth", "_nginx"]
+ALLOWED_ROUTES = os.environ.get('TURNPIKE_ALLOWED_ROUTES', ["public", "api", "app"])
+ALLOWED_NO_AUTH_ROUTES = os.environ.get('TURNPIKE_NO_AUTH_ROUTES', ["public"])
+ALLOWED_ORIGIN_DOMAINS = os.environ.get('TURNPIKE_ALLOWED_ORIGIN_DOMAINS', [".svc.cluster.local"])
+
+
+def validate_route(backend):
```
[the remark this answers is not in the document]
Comment (j00bar) [reply nginx/build_config.py:22]: 😬 Dammit.

File: nginx/build_config.py [hunk 2]
```diff
@@ -13,7 +13,38 @@
 
 
 # Do not include leading or trailing slashes in these routes
-FORBIDDEN_ROUTES = ["", "saml", "auth", "_nginx"]
+PROTECTED_ROUTES = ["saml", "auth", "_nginx"]
+ALLOWED_ROUTES = os.environ.get('TURNPIKE_ALLOWED_ROUTES', ["public", "api", "app"])
+ALLOWED_NO_AUTH_ROUTES = os.environ.get('TURNPIKE_NO_AUTH_ROUTES', ["public"])
```
[the remark this answers is not in the document]
Comment (j00bar) [reply nginx/build_config.py:18]: I wanted to at least code for the possibility. We can disable using environment variables if we want.

[the remark this answers is not in the document]
Comment (coderbydesign) [reply nginx/build_config.py:18]: 🚀

**Review by coderbydesign — APPROVED** [review]
This looks great! 🎉

File: nginx/build_config.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+import warnings
+
+import jinja2
+import yaml
+import yaml.error
+
+
+# Do not include leading or trailing slashes in these routes
+PROTECTED_ROUTES = ["saml", "auth", "_nginx"]
+ALLOWED_ROUTES = json.loads(os.environ.get('TURNPIKE_ALLOWED_ROUTES', '["public", "api", "app"]'))
```
Comment (coderbydesign) [inline nginx/build_config.py:17]: Is the plan to set `TURNPIKE_ALLOWED_ROUTES` in prod to allow for `docs`?

Comment (j00bar) [reply nginx/build_config.py:17]: Yep!

**Pull request merged**